SputnikVM/EVM-RS integration

mersinvald opened the pull request:
This PR proposes inclusion of [EVM-RS]() (through [EVM-FFI]()) as an Ethereum Virtual Machine backend for multi-geth

Integration implements the following properties:
 - opt-in at build time with `--tags sputnikvm` flag
 - opt-in at runtime with `--vm.evm=svm` flag
 - supports only x86_64 and i686 (for now)
 - no support for cross-compilation

The described limitations are temporary and are guaranteed to be held by build scripts and multi-geth internal checks.

whilei committed:
feat: evm-ffi integration for evm-rs support as an evm backend

Commit:
chores: switch to ethereumproject/evm-ffi

tzdybal committed:
vendor: evm-ffi added

tzdybal committed:
build.sh: small bugfixes

Fixed:
 - `ethereumproject` added to paths
 - `GOCACHE` removed, because of Go 1.12 compatibility issues
 - relative path from build.sh to vendor, for simplicity

Commit:
evm-rs: EVM-RS support with a dedicated evm-rs build script, architecture and cross-build detection

Commit:
feat: fail fast if --vm.evm=svm and geth wasn't build with evm-rs support

meowsbits committed:
build/env.sh: fixes 'unexpected operator warn/error'

$ make geth
build/env.sh go run build/ci.go install ./cmd/geth
build/env.sh: 30: [: false: unexpected operator
>>> /home/ia/go1.12.1.linux-amd64/go/bin/go install -ldflags -X main.gitCommit=bc0dcbfe1ac2d53d45888fbb0d03444ad3f60f79 -X main.gitDate=20190520 -v ./cmd/geth
Done building.
Run /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/geth to launch geth.

meowsbits committed:
build/env.sh: use explicit '*' instead of implicit arr concat

re: SC2199 linter warning

   SC2199 Arrays implicitly concatenate in [[ ]]. Use a loop (or explicit * instead of @). (sh-shellcheck)

meowsbits committed:
build/env.sh: in POSIX sh, source as alias for . is undef

meowsbits committed:
build/clean_evm-rs.sh: nitpickery

meowsbits commented on core/state_processor_svm.go:
```diff
@@ -0,0 +1,326 @@
+		HasReducedSstoreGasMetering: rules.IsEIP1283F,
+		ErrOnCallWithMoreGas:        !rules.IsEIP150,
+		CallCreateL64AfterGas:       rules.IsEIP150,
+		MemoryLimit:                 ^uint(0), // Reversed 0 is max unsigned integer value for uint
```
:white_flower:

mersinvald replied:
@meowsbits afair I couldn’t find any parameter that defines memory limit, and the classic geth w/ SVM was operating the same way, with an unlimited memory storage. Though that’s pretty much welcoming for a DOS attack :)

Is there any limit definition I could use?

meowsbits replied:
https://github.com/multi-geth/multi-geth/blob/master/core/vm/memory_table.go
https://github.com/multi-geth/multi-geth/blob/master/core/vm/common.go#L38

[18446744073709551615](https://docs.microsoft.com/en-us/dotnet/api/system.uint64.maxvalue?view=netframework-4.8)

```go
$ cat max-uint64.go 
package main

import (
        "fmt"
        "log"
        "math/big"
)

func main() {
        maybeMax := new(big.Int).SetUint64(18446744073709551615)
        b := new(big.Int).Sub(maybeMax, big.NewInt(10))
        for {
                b = b.Add(b, big.NewInt(1))
                if !b.IsUint64() {
                        log.Fatal("max uint64 is", b, "-1")
                } else {
                        fmt.Println(b)
                }
        }
}

```

meowsbits committed:
internal/build: gofmt

tzdybal commented on build/ci.go:
```diff
@@ -282,17 +282,24 @@ func doInstall(cmdline []string) {
 
 func buildFlags(env build.Environment) (flags []string) {
 	var ld []string
+
```
This is quite obvious, but we will have to take care about this during every rebase with upstream.

tzdybal commented on build/env.sh:
```diff
@@ -26,5 +26,15 @@ export GOPATH
+        echo "Cross-builds are not yet supported with EVM-RS enabled"
+        exit 1
+    fi
+    . build/build_evm-rs.sh
```
Use `source` instead of  `.`, for POSIX sh compatibility.

meowsbits replied:
Wait a second...

bb2d121 

![Screenshot_20190521_075500](https://user-images.githubusercontent.com/45600330/58097776-c4011d00-7b9d-11e9-8efd-d3cb26a22661.png)

tzdybal replied:
I read this error message the wrong way ;)

tzdybal commented on build_with_svm.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+pushd ./vendor/github.com/ethereumproject/evm-ffi/c/
```
Or simply:
```
make -C ./vendor/github.com/ethereumproject/evm-ffi/c/
```
without `pushd`/`popd`.

tzdybal commented on core/state_transition.go:
```diff
@@ -197,7 +197,6 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
 	if err = st.useGas(gas); err != nil {
 		return nil, 0, false, err
 	}
-
```
Again, this is a simple change, but can cause some extra work during merge with upstream. Or I'm missing something? (CC: @meowsbits)

mersinvald replied:
Seems to be fixed by @meowsbits

tzdybal commented on test_with_svm.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+pushd ./vendor/github.com/ethereumproject/evm-ffi/c/
```
`make -C ./vendor/github.com/ethereumproject/evm-ffi/c/` instead of `pushd`/`popd`

tzdybal reviewed (commented):
I found some little issues. Most important is probably usage of `.` instead of `source`.

Commit:
fix(scripts): pushd/popd -> simple make

Commit:
fix: remove test_with_svm.sh as it's misleading: it won't run tests with SVM as current test harness doesn't support choosing the evm.

Commit:
fix(style): remove unnecessary linebreak

meowsbits commented on build/build_evm-rs.sh:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
```
here?

meowsbits commented on core/state_processor_svm.go:
```diff
@@ -0,0 +1,326 @@
+
+	// Set the receipt logs and create a bloom for filtering
+	receipt.Logs = statedb.GetLogs(tx.Hash())
+	receipt.Bloom = types.CreateBloom(types.Receipts{receipt})
```
```suggestion
	receipt.Bloom = types.CreateBloom(types.Receipts{receipt})
	receipt.BlockHash = statedb.BlockHash()
	receipt.BlockNumber = header.Number
	receipt.TransactionIndex = uint(statedb.TxIndex())

```

meowsbits replied:
resolved w/ https://github.com/multi-geth/multi-geth/pull/82/commits/17d8f006e98f3e0dfe42660aefc0af8442aecaa6

meowsbits committed:
build: fix compilation issue

https://github.com/multi-geth/multi-geth/pull/82#issuecomment-499992057

Use user's env bash specifically.

meowsbits committed:
core: ApplySputnikTransaction should set moar receipt vals

The setting of these values corresponds to the behavior at
https://github.com/multi-geth/multi-geth/blob/master/core/state_processor.go#L124-L126

meowsbits committed:
tests,eth/tracers: enable tests suite to run with svm config

Duplicating the MakePreState fn from exported val in tests/
to eth/tracers resolves an circular import problem that happens
given moving the tests/ suite testVMConfig value from _test.go
file to package level at init.go

This also modifies the withTrace fn to carry over
vm.config interpreter preferences.
